runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/jobsubmission/JobServerDriver.java

```diff
@@ -0,0 +1,234 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.runners.fnexecution.jobsubmission;
+
+import java.io.IOException;
+import java.nio.file.Paths;
+import java.util.function.Function;
+import org.apache.beam.model.pipeline.v1.Endpoints;
+import org.apache.beam.runners.fnexecution.GrpcFnServer;
+import org.apache.beam.runners.fnexecution.ServerFactory;
+import org.apache.beam.runners.fnexecution.artifact.BeamFileSystemArtifactStagingService;
+import org.apache.beam.vendor.guava.v20_0.com.google.common.annotations.VisibleForTesting;
+import org.kohsuke.args4j.Option;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** Shared code for starting and serving an {@link InMemoryJobService}. */
+public abstract class JobServerDriver implements Runnable {
+
+  protected abstract JobInvoker createJobInvoker();
+
+  private static final Logger LOG = LoggerFactory.getLogger(JobServerDriver.class);
+
+  @VisibleForTesting public ServerConfiguration configuration;
+
+  private final ServerFactory jobServerFactory;
+  private final ServerFactory artifactServerFactory;
+  private volatile GrpcFnServer<InMemoryJobService> jobServer;
+  private volatile GrpcFnServer<BeamFileSystemArtifactStagingService> artifactStagingServer;
+
+  /** Configuration for the jobServer. */
+  public static class ServerConfiguration {
+    @Option(name = "--job-host", usage = "The job server host name")
+    private String host = "localhost";
+
+    @Option(
+        name = "--job-port",
+        usage = "The job service port. 0 to use a dynamic port. (Default: 8099)")
+    private int port = 8099;
+
+    @Option(
+        name = "--artifact-port",
+        usage = "The artifact service port. 0 to use a dynamic port. (Default: 8098)")
+    private int artifactPort = 8098;
+
+    @Option(name = "--artifacts-dir", usage = "The location to store staged artifact files")
+    private String artifactStagingPath =
+        Paths.get(System.getProperty("java.io.tmpdir"), "beam-artifact-staging").toString();
+
+    @Option(
+        name = "--clean-artifacts-per-job",
+        usage = "When true, remove each job's staged artifacts when it completes")
+    private boolean cleanArtifactsPerJob = false;
+
+    @Option(
+        name = "--sdk-worker-parallelism",
+        usage = "Default parallelism for SDK worker processes (see portable pipeline options)")
+    private Long sdkWorkerParallelism = 1L;
+
+    @Option(
+        name = "--artifact-service-enabled",
+        usage = "When false, the artifact staging service will not be started and set to null")
+    private boolean artifactServiceEnabled = true;
```

Comment:
Lets add test for `artifactServiceEnabled = false`

Reply:
Done.

Reply:
turns out CmdLineParser doesn't let you set boolean values (?) so I changed it to --artifact-service-disabled

Reply:
I guess you can set `ExplicitBooleanOptionHandler` but I don't really care either way :man_shrugging: